### PR TITLE
Cryptopkg provide SM3-only OpensslLib for MbedTlsLib

### PR DESCRIPTION
**Pull request by ardbiesheuvel**
Provide an OpensslLib variant that carries only the SM3 code, which is all MbedTlsLib needs from it. This removes the need for providing OpenSSL on 32-bit ARM, which results in the need for a softfloat library.

Other explored solutions (#6165, #6189)

- Add new SM3 implementation into BaseCryptLibMbedTls (rejected by @liyi77)
- Copy SM3 code from OpenSSL into BaseCryptLibMbedTls (rejected by @mdkinney)
- Contribute SM3 support to MbedTLS - existing effort stalled for 2 years, and outdated now; MbedTLS API itself has evolved and top-of-tree can no longer be used with EDK2's existing glue code.
- Remove SM3 support from the 32-bit ARM build of MbedTlsLib (rejected by @leiflindholm on the basis that all library definitions now need to handle ARM in a special way)
- Reference the sm3.c source file in the OpensslLib directory from BaseCryptLibMbedTLs
- Reference the sm3.c source file in the OpensslLib directory from MbedTlsLib

### CODE REVIEW

**Comment by ardbiesheuvel** [comment]
cc @os-d

**Review by os-d — APPROVED** [review]

**Comment by mdkinney** [comment]
Does the title and description of this PR need to be updated?  I do not see sm3.c in this PR anymore.

**Comment by ardbiesheuvel** [comment]
> Does the title and description of this PR need to be updated? I do not see sm3.c in this PR anymore.

No. The series essentially replaces

```
[LibraryClasses]
  OpensslLib
```

with

```
  $(OPENSSL_PATH)/crypto/sm3.
```

which is reasonable IMHO because the OpenSSL submodule is part of CryptoPkg.

**Review by kraxel — APPROVED** [review]

**Comment by ardbiesheuvel** [comment]
@liyi77 @jyao1 any thoughts regarding this approach?

**Comment by ardbiesheuvel** [comment]
@mdkinney Does this look ok to you now?

**Comment by liyi77** [comment]
I have no problem with reusing sm3 of openssl, but I remember Mike saying he didn't want relative paths in the INF.

**Comment by ardbiesheuvel** [comment]
> I have no problem with reusing sm3 of openssl, but I remember Mike saying he didn't want relative paths in the INF.

OK, so given that the OpenSSL source is shared between OpensslLib and MbedTlsLib, we'll need to move the Git submodule to CryptoPkg/Library/ then

File: CryptoPkg/Library/BaseCryptLibMbedTls/BaseCryptLib.inf [hunk 1]
```diff
@@ -58,9 +60,10 @@
 
   SysCall/CrtWrapper.c
   SysCall/DummyOpensslSupport.c
-  SysCall/BaseMemAllocation.c
   SysCall/TimerWrapper.c
 
+  $(OPENSSL_PATH)/crypto/sm3/sm3.c
```
Comment (liyi77) [inline CryptoPkg/Library/BaseCryptLibMbedTls/BaseCryptLib.inf:65]: Could add sm3 file in MbedtlsLib? Then you don't have to add it to every CryptLib.inf

Comment (ardbiesheuvel) [reply CryptoPkg/Library/BaseCryptLibMbedTls/BaseCryptLib.inf]: Done

**Comment by jyao1** [comment]
I have no objection from functional perspective.

I would rely on @mdkinney 's comment on EDKII INF rule.

**Comment by ardbiesheuvel** [comment]
> I have no objection from functional perspective.
> 
> I would rely on @mdkinney 's comment on EDKII INF rule.

OK. I will go with the following instead

```diff
--- a/CryptoPkg/Library/MbedTlsLib/MbedTlsLib.inf
+++ b/CryptoPkg/Library/MbedTlsLib/MbedTlsLib.inf
@@ -14,7 +14,7 @@ [Defines]
   VERSION_STRING                 = 1.0
   LIBRARY_CLASS                  = MbedTlsLib
 
-  DEFINE OPENSSL_PATH            = ../OpensslLib/openssl
+  DEFINE OPENSSL_PATH            = CryptoPkg/Library/OpensslLib/openssl
 
 #
 # The following information is for reference only and not required by the build tools.
```

**Comment by ardbiesheuvel** [comment]
> > I have no objection from functional perspective.
> > I would rely on @mdkinney 's comment on EDKII INF rule.
> 
> OK. I will go with the following instead
> 
> ```diff
> --- a/CryptoPkg/Library/MbedTlsLib/MbedTlsLib.inf
> +++ b/CryptoPkg/Library/MbedTlsLib/MbedTlsLib.inf
> @@ -14,7 +14,7 @@ [Defines]
>    VERSION_STRING                 = 1.0
>    LIBRARY_CLASS                  = MbedTlsLib
>  
> -  DEFINE OPENSSL_PATH            = ../OpensslLib/openssl
> +  DEFINE OPENSSL_PATH            = CryptoPkg/Library/OpensslLib/openssl
>  
>  #
>  # The following information is for reference only and not required by the build tools.
> ```

This builds on Linux but not on Windows. I will try to think of another way to approach this.

**Comment by ardbiesheuvel** [comment]
@liyi77 @jyao1 

Attempt 7: provide a OpenssLib version that has only the SM3 code and nothing else. Please let me know what you think.

**Review by liyi77 — APPROVED** [review]

**Comment by liyi77** [comment]
> @liyi77 @jyao1
> 
> Attempt 7: provide a OpenssLib version that has only the SM3 code and nothing else. Please let me know what you think.

Looks good.

**Pull request merged**